models/aeif_psc_delta.cpp

```diff
@@ -0,0 +1,526 @@
+/*
+ *  aeif_psc_delta.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "aeif_psc_delta.h"
+
+#ifdef HAVE_GSL
+
+// C++ includes:
+#include <cmath>
+#include <cstdio>
+#include <iomanip>
+#include <iostream>
+#include <limits>
+
+// Includes from libnestutil:
+#include "numerics.h"
+
+// Includes from nestkernel:
+#include "exceptions.h"
+#include "kernel_manager.h"
+#include "nest_names.h"
+#include "universal_data_logger_impl.h"
+
+// Includes from sli:
+#include "dict.h"
+#include "dictutils.h"
+#include "doubledatum.h"
+#include "integerdatum.h"
+
+/* ----------------------------------------------------------------
+ * Recordables map
+ * ---------------------------------------------------------------- */
+
+nest::RecordablesMap< nest::aeif_psc_delta >
+  nest::aeif_psc_delta::recordablesMap_;
+
+namespace nest
+{
+/*
+ * template specialization must be placed in namespace
+ *
+ * Override the create() method with one call to RecordablesMap::insert_()
+ * for each quantity to be recorded.
+ */
+template <>
+void
+RecordablesMap< aeif_psc_delta >::create()
+{
+  // use standard names whereever you can for consistency!
+  insert_(
+    names::V_m, &aeif_psc_delta::get_y_elem_< aeif_psc_delta::State_::V_M > );
+  insert_(
+    names::w, &aeif_psc_delta::get_y_elem_< aeif_psc_delta::State_::W > );
+}
+}
+
+
+extern "C" int
+nest::aeif_psc_delta_dynamics( double,
+  const double y[],
+  double f[],
+  void* pnode )
+{
+  // a shorthand
+  typedef nest::aeif_psc_delta::State_ S;
+
+  // get access to node so we can almost work as in a member function
+  assert( pnode );
+  const nest::aeif_psc_delta& node =
+    *( reinterpret_cast< nest::aeif_psc_delta* >( pnode ) );
+
+  // y[] here is---and must be---the state vector supplied by the integrator,
+  // not the state vector in the node, node.S_.y[].
+
+  // The following code is verbose for the sake of clarity. We assume that a
+  // good compiler will optimize the verbosity away ...
+
+  // shorthand for state variables
+  const double& V = std::min( y[ S::V_M ], node.P_.V_peak_ ); // bind V to the
```

Comment:
This should not be a reference, since `std::min()` might return a temporary value.